### PR TITLE
Fixes Issue #35429 - unmatched tcp traffic external name service

### PR DESCRIPTION
**Pull request by eliavem**
**Please provide a description of this PR:**
This PR fixes issue #35429. It adds an error issue for ExternalName service type when service does not follow naming convention or has an unnamed port

### CODE REVIEW

File: pkg/config/analysis/msg/messages.yaml [hunk 1]
```diff
@@ -548,3 +548,10 @@ messages:
         type: string
       - name: pod
         type: string
+        
+  - name: "ExternalNameServiceTypeInvalidPortName"
```
[the remark this answers is not in the document]
Comment (linsun) [reply pkg/config/analysis/msg/messages.yaml:552]: with the existing `ServiceEntryAddressesRequired`, seems there is still value to add this warning msg as well?

[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/config/analysis/msg/messages.yaml:552]: I agree, I just don't think that its sufficient. It doesn't cover the case of `tcp` named port. I think we should use the same logic as `ExternalNameServiceTypeInvalidPortName` - we can give it a new message if want

[the remark this answers is not in the document]
Comment (eliavem) [reply pkg/config/analysis/msg/messages.yaml:552]: @howardjohn what if I modify the issue name from ExternalNameServiceTypeInvalidPortName to something that would cover for ports named as "tcp" as well and change the message?

In port name analyzer, I can add a check for external names service type + tcp port names. Then, we can display warning for issue IST0150 when any of the following is true:

 1. We have an external name svc + unnamed port
 2. We have an external name svc + invalid name port 
 3. We have an external name svc + tcp named port

[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/config/analysis/msg/messages.yaml:552]: :+1:

[the remark this answers is not in the document]
Comment (linsun) [reply pkg/config/analysis/msg/messages.yaml:552]: SGTM, thanks @eliavem and @howardjohn !

**Comment by Shegox** [comment]
@googlebot rescan

**Comment by eliavem** [comment]
@linsun I'm part of a google group now. Can we do a "Re-run" on this PR?

**Comment by linsun** [comment]
@eliavem probably worth adding a note in the release note dir since this is user facing.

File: releasenotes/notes/35429.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: telemetry
+issue:
+  - 35429
+
+releaseNotes:
+- |
+  **Added** `istioctl analyze` will display a warning when tcp service of type ServiceName have invalid port name or name is tcp.
```
Comment (linsun) [inline releasenotes/notes/35429.yaml:9]:
```suggestion
  **Added** `istioctl analyze` will display a warning when tcp service of type ExternalName have invalid port name or name is tcp.
```

**Comment by linsun** [comment]
LGTM @eliavem, just one minor suggestion

**Comment by eliavem** [comment]
@howardjohn @linsun could someone take a look?

File: releasenotes/notes/35429.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: telemetry
```
Comment (bianpengyuan) [inline releasenotes/notes/35429.yaml:3]: This is not telemetry area. should be istioctl

File: releasenotes/notes/35429.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+issue:
+  - 35429
+
+releaseNotes:
+- |
+  **Added** `istioctl analyze` will display a warning when tcp service of type ExternalName have invalid port name or name is tcp.
```
Comment (bianpengyuan) [inline releasenotes/notes/35429.yaml:9]:
```suggestion
  **Added** `istioctl analyze` will display a warning when service of type ExternalName have invalid port name or name is tcp.
```

File: pkg/config/analysis/analyzers/service/portname.go [hunk 5]
```diff
@@ -67,15 +67,20 @@ func (s *PortNameAnalyzer) Analyze(c analysis.Context) {
...
-		if instance := configKube.ConvertProtocol(port.Port, port.Name, port.Protocol, port.AppProtocol); instance.IsUnsupported() {
+		instance := configKube.ConvertProtocol(port.Port, port.Name, port.Protocol, port.AppProtocol)
+		if instance.IsUnsupported() || port.Name == "tcp" && svc.Type == "ExternalName" {
 
 			m := msg.NewPortNameIsNotUnderNamingConvention(
 				r, port.Name, int(port.Port), port.TargetPort.String())
 
+			if svc.Type == "ExternalName" {
+				m = msg.NewExternalNameServiceTypeInvalidPortName(
+					r)
```
Comment (bianpengyuan) [inline pkg/config/analysis/analyzers/service/portname.go:78]:
```suggestion
				m = msg.NewExternalNameServiceTypeInvalidPortName(r)
```

**Review by bianpengyuan — APPROVED** [review]

**Pull request merged**